### PR TITLE
Antibody Modeling (WIP)

### PR DESCRIPTION
**Pull request by dhuvik**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by dhuvik** [comment]
Working on Unit Tests that Test both CPU and GPU for hardware accelerators.

File: deepchem/models/torch_models/antibody_modeling.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
+from transformers import AutoModel, AutoModelForMaskedLM, AutoTokenizer, AutoConfig
+from tqdm import tqdm
+from deepchem.models.torch_models import HuggingFaceModel
+from typing import Union
+
+
+class DeepAbLLM(HuggingFaceModel):
```
Comment (rbharath) [inline deepchem/models/torch_models/antibody_modeling.py:7]: Need to add this to the RST docs

File: deepchem/models/torch_models/antibody_modeling.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    """Flexible Antibody Language Model for Re-Design of Ab Residues.
+
+    This wrapper class is designed to inherit from the HuggingFaceModel
+    object from the DeepChem repositor. The class is designed to leverage
```
Comment (rbharath) [inline deepchem/models/torch_models/antibody_modeling.py:11]: Describe what the class actually does; less so what it is.

File: deepchem/models/torch_models/antibody_modeling.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+            ' ', '')  # Get ride of extraneous spaces if any
+        cleaned_sequence = list(
+            cleaned_sequence)  # Turn the sequence into a list
+        cleaned_sequence[idx] = '*'  # Mask the sequence at idx
+        masked_sequence = ' '.join(cleaned_sequence)  # Convert list -> seq
+        masked_sequence = masked_sequence.replace('*',
+                                                  self.tokenizer.mask_token)
+        return masked_sequence
+
+    def redesign_residue(self,
```
Comment (rbharath) [inline deepchem/models/torch_models/antibody_modeling.py:148]: We need unit tests

**Review by rbharath — COMMENTED** [review]
On the right track in this preliminary review; need some unit tests and doc fixes

File: deepchem/models/torch_models/antibody_modeling.py (hunk 1 above)
Comment (dhuvik) [reply deepchem/models/torch_models/antibody_modeling.py:7]: Added to models.rst

File: deepchem/models/torch_models/antibody_modeling.py (hunk 3 above)
Comment (dhuvik) [reply deepchem/models/torch_models/antibody_modeling.py:147]: Unit tests and integration tests have added

File: deepchem/models/torch_models/antibody_modeling.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        Number of tasks for a given model. Default: 1
+
+    Methods
+    -------
+    __init__(task, model_path, n_tasks)
+        Initialize an DeepAbLLM with specified information.
+    _mask_seq_pos(position, idx)
+        Mask a sequence at a particular index.
+    redesign_residue(n=1.0)
+        Change the photo's gamma exposure.
```
Comment (aaronrockmenezes) [inline deepchem/models/torch_models/antibody_modeling.py:35]: Change docs

File: deepchem/models/torch_models/antibody_modeling.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    optimize_sequence(n=1.0)
+        Change the photo's gamma exposure.
+
+    Notes
+    -----
+    Model currently supports:
+    1. Light or Heavy Chain Re-Design at Arbitrary Point
+
+    (WIP)
```
Comment (aaronrockmenezes) [inline deepchem/models/torch_models/antibody_modeling.py:46]: Update

File: deepchem/models/torch_models/antibody_modeling.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+                 n_tasks: int = 1,
+                 **kwargs) -> None:
+        """
+        Parameters
+        ----------
+        task: str
+            The task defines the type of learning task in the model. The supported tasks are
+            - `mlm` - masked language modeling commonly used in pretraining
+            - `classification` - use it for classification tasks
+        model_path: str
```
Comment (aaronrockmenezes) [inline deepchem/models/torch_models/antibody_modeling.py:92]: Change to hf_model_link/hf_model_path

File: deepchem/models/torch_models/antibody_modeling.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        Parameters
+        ----------
+        sequence: str
+            The antibody sequence to redesign.
+        residue_index: int
+            The residue index to mask and unmask.
+        top_k: int
+            The top_k logits to return. Defaults to 10.
+        verbose: bool
+            If verbose, prints the original sequence and the resiude at residue_index
```
Comment (aaronrockmenezes) [inline deepchem/models/torch_models/antibody_modeling.py:165]: spell check residue

File: deepchem/models/torch_models/antibody_modeling.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+                             result.get('score', None))
+                            for result in results
+                            if isinstance(result, dict)]]
+        return sequence_tuples
+
+    def optimize_residue_pos(self,
+                             sequence: str,
+                             residue_index: int,
+                             verbose: bool = False,
+                             **kwargs):
```
Comment (aaronrockmenezes) [inline deepchem/models/torch_models/antibody_modeling.py:195]: Pass threshold directly with default to 0.0 instead of kwarg

**Review by aaronrockmenezes — COMMENTED** [review]
Added a few comments. No important changes needed

**Comment by aaronrockmenezes** [comment]
LGTM once tests pass

File: deepchem/models/torch_models/antibody_modeling.py (hunk 5 above)
Comment (dhuvik) [reply deepchem/models/torch_models/antibody_modeling.py:46]: Updated

File: deepchem/models/torch_models/antibody_modeling.py (hunk 4 above)
Comment (dhuvik) [reply deepchem/models/torch_models/antibody_modeling.py:35]: Doc strings are fixed and typo free

File: deepchem/models/torch_models/antibody_modeling.py (hunk 6 above)
Comment (dhuvik) [reply deepchem/models/torch_models/antibody_modeling.py:92]: On second read: model_path accepts local or HF paths as inputs so I think this ones okay as is.

File: deepchem/models/torch_models/antibody_modeling.py (hunk 7 above)
Comment (dhuvik) [reply deepchem/models/torch_models/antibody_modeling.py:165]: Thank you for the catch!

**Comment by dhuvik** [comment]
Currently not faililng any tests due to the code added in the PR.

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**